### PR TITLE
CI: Update Linux platforms / runners, remove `ubuntu-trusty`, `opensuse-{15.3,15.4}`

### PR DESCRIPTION
**Pull request by mkoeppe**
We increase parallelization according to [the upgraded Linux runners on GH Actions](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) (4 CPUs) and add some newer platforms.

`fedora-40` fails because we reject GCC 14, that's now tracked in #37353.

Update (Apr 2024): 
- ubuntu-trusty reaches EOL, we drop it. Also remove some past-EOL OpenSUSE Leap; see #32074
- also reducing the list of systems tested with `--enable-system-site-packages` because there is no interest in doing the porting work
- updating the default platform (used by the incremental build workflows) from `ubuntu-focal` to `ubuntu-jammy`.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -93,16 +93,16 @@ jobs:
           ./bootstrap && make build
         working-directory: ./worktree-image
         env:
-          MAKE: make -j2 --output-sync=recurse
-          SAGE_NUM_THREADS: 2
+          MAKE: make -j4 --output-sync=recurse
```
Comment (tobiasdiez) [inline .github/workflows/build.yml:96]: Did github changed to 4 cores now?

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -25,6 +25,7 @@ on:
            "ubuntu-jammy",
            "ubuntu-lunar",
            "ubuntu-mantic",
+           "ubuntu-noble",
```
Comment (tobiasdiez) [inline .github/workflows/docker.yml:27]: Please remove some old systems to keep the maintenance effort and the cpu time (relatively) constant.

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -46,13 +48,15 @@ on:
            "fedora-37",
            "fedora-38",
            "fedora-39",
+           "fedora-40",
```
Comment (tobiasdiez) [inline .github/workflows/docker.yml:50]: Please remove fedora-40 until it is green.

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
Please reduce the scope before applying the blocker label.

**Comment by mkoeppe** [comment]
Stop this nonsense, Tobias.

**Comment by tobiasdiez** [comment]
Removing blocker as there are problems with other PRs, see https://github.com/sagemath/sage/issues/37428 for more details.

**Comment by mkoeppe** [comment]
> Removing blocker as there are problems with other PRs, see #37428 for more details.

That PR has absolutely no details. 

Mass removal of the "blocker" label on 9 PRs is not OK. @roed314 @jhpalmieri @williamstein 

And even when removing "blocker" is necessary in a case to fix a problem that it causes on other PRs, the mildest form of label change should be used -- reducing to "critical".

File: .github/workflows/docker.yml (hunk 2 above)
Comment (mkoeppe) [reply .github/workflows/docker.yml:27]: No, we will continue to test what is supported.

File: .github/workflows/docker.yml (hunk 3 above)
Comment (mkoeppe) [reply .github/workflows/docker.yml:50]: Platforms don't magically become green. Developers who work on the portability look at such test runs to find out what needs to be fixed.

**Comment by mkoeppe** [comment]
@vbraun please merge

**Comment by vbraun** [comment]
I don't see anything here that should be a blocker

**Comment by mkoeppe** [comment]
@vbraun Just a bit hard to see whether the release is ready when the CI is broken, that's all.

File: .github/workflows/ci-linux-incremental.yml [hunk 4]
```diff
@@ -123,8 +123,7 @@ jobs:
       # Only test systems with a usable system python (>= 3.9)
       # with only a small number of test failures as of 10.2.rc0
       tox_system_factors: >-
-        ["gentoo-python3.11",
-         "archlinux-latest"]
+        ["gentoo-python3.11"]
       tox_packages_factors: >-
```
Comment (kwankyu) [inline .github/workflows/ci-linux-incremental.yml]: Why deleted "archlinux-latest"?

Comment (mkoeppe) [reply .github/workflows/ci-linux-incremental.yml]: it was failing, and there seems to be nobody who's interested in fixing the "sitepackages" mode on platforms that they don't use themselves

Comment (kwankyu) [reply .github/workflows/ci-linux-incremental.yml]: "archlinux-latest" is one of the few systems that succeeded in CI linux run for 10.3.beta8 https://github.com/sagemath/sage/actions/runs/7894201915 ?

Comment (mkoeppe) [reply .github/workflows/ci-linux-incremental.yml]: Indeed it did. OK, I'll take out this change.

Comment (kwankyu) [reply .github/workflows/ci-linux-incremental.yml]: Thanks.

**Review by kwankyu — APPROVED** [review]
LGTM. This PR just expands the testing facility to more systems.

**Comment by mkoeppe** [comment]
Thank you!

**Comment by mkoeppe** [comment]
@vbraun Positively reviewed, please merge. The rest = label manipulation by the usual obstructionists.

**Comment by roed314** [comment]
This PR was given a positive review by @kwankyu, and @tobiasdiez undoing that review is not in accordance with Sage's reviewing practices.  If the vote [here](https://groups.google.com/g/sage-devel/c/IgBYUJl33SQ) succeeds, those involved may act in accordance with the new policy.  In the meantime I'm setting this to positive review.

Given the status of CI and the amount of time that this ticket has been stalled, updates may need to be made.  I ask those involved with this ticket to take reasonable actions in preparation for it getting merged, and @vbraun to give them sufficient time to do so.

We're leaving this at critical for now, but encourage those involved to discuss whether it should be set to blocker/CI Fix.

**Comment by jhpalmieri** [comment]
I'm posting to record a vote of -1 on behalf of Tobias Diez.

**Comment by kwankyu** [comment]
Still +1 as this PR is a clear improvement.

**Comment by roed314** [comment]
Based on @kwankyu's [suggestion](https://groups.google.com/g/sage-devel/c/IgBYUJl33SQ/m/4Hls5k3MCgAJ) and discussion by the new Code of Conduct Committee, I'm setting this ticket to Needs Review so that all the disputed tickets start with a clean slate prior to being update based on voting.  Starting at midnight US/Pacific time tonight anyone involved should feel free to set the status to positive review or needs review in accordance with the [new policy](https://groups.google.com/g/sage-devel/c/IgBYUJl33SQ/m/kvmOlVb1AQAJ).

**Comment by roed314** [comment]
I'm recording a -1 from Dima Pasechnik.

File: .github/workflows/docker.yml (hunk 2 above)
Comment (saraedum) [reply .github/workflows/docker.yml:27]: Do you mean supported by SageMath or the Ubuntu organization?

Comment (saraedum) [reply .github/workflows/docker.yml:27]: image

If the latter, I am not sure we should worry about "pro support". But maybe we have a policy about that that I am not aware of?

File: .github/workflows/build.yml (hunk 1 above)
Comment (saraedum) [reply .github/workflows/build.yml:96]: Yes, see https://github.blog/2024-01-17-github-hosted-runners-double-the-power-for-open-source/

**Comment by saraedum** [comment]
:+1: from me. I don't see a problem with this PR.

File: .github/workflows/docker.yml (hunk 2 above)
Comment (mkoeppe) [reply .github/workflows/docker.yml:27]: @saraedum I mean supported by SageMath.

Comment (mkoeppe) [reply .github/workflows/docker.yml:27]: @saraedum for reference, what is supported by SageMath and why we drop support is tracked in #32074.

**Comment by mkoeppe** [comment]
> However, I don't understand what was "disputed" about this PR.

Everything you need to know about these "disputes" about the CI runs:
- https://github.com/sagemath/sage/pull/36726#issuecomment-1847870000
- https://github.com/sagemath/sage/pull/36726#issuecomment-1872331240

**Comment by saraedum** [comment]
> > However, I don't understand what was "disputed" about this PR.
> 
> Everything you need to know about these "disputes" about the CI runs:
> 
>     * [CI Linux: Replace use of pkill #36726 (comment)](https://github.com/sagemath/sage/pull/36726#issuecomment-1847870000)
> 
>     * [CI Linux: Replace use of pkill #36726 (comment)](https://github.com/sagemath/sage/pull/36726#issuecomment-1872331240)

Sorry, I fail to understand what's the relation of these posts with the reservations that they might have with this PR.

File: .github/workflows/ci-linux.yml [hunk 5]
```diff
@@ -111,18 +111,11 @@ jobs:
         ["ubuntu-jammy",
          "ubuntu-lunar",
          "ubuntu-mantic",
-         "debian-bullseye",
          "debian-bookworm",
```
Comment (mkoeppe) [inline .github/workflows/ci-linux.yml:114]: @orlitzky The changes in this file reduce the platforms tested with `--enable-system-site-packages`. Is this change OK with you? 
(See https://github.com/sagemath/sage/actions/runs/8501570896 for the tests with 10.4.beta1)

Comment (orlitzky) [reply .github/workflows/ci-linux.yml:114]: It makes no difference to me. Having e.g. debian-bullseye in the CI is useful only to people who want to use system packages on Debian bullseye. I don't think we have any -- not developers, anyway.

**Comment by tornaria** [comment]
> [Documentation preview for this PR](https://deploy-preview-37351--sagemath.netlify.app/html/en) (built with commit [bce5774](https://github.com/sagemath/sage/commit/bce577499188a4d519eb79de518b690e28a88d58); [changes](https://deploy-preview-37351--sagemath.netlify.app/CHANGES.html)) is ready! 🎉

If you are annoyed by this useless bot spam, please :+1: on #37740.

**Comment by mkoeppe** [comment]
Let's get this in please.

@saraedum Not sure if this should still be "disputed". By passage of time, we are targeting a different release, and I've made numerous changes to the branch since.

**Comment by saraedum** [comment]
@mkoeppe I think this is currently how our rules work. If a PR is disputed, it remains disputed. That's maybe something to change but I think this is not for this PR to discuss.

## Current Votes (2-2)

:+1: @mkoeppe (author)
:+1: @kwankyu https://github.com/sagemath/sage/pull/37351#issuecomment-1998614861
:-1: @tobiasdiez https://github.com/sagemath/sage/pull/37351#issuecomment-1998385430
:-1: @dimpase https://github.com/sagemath/sage/pull/37351#issuecomment-2000734246

**Comment by roed314** [comment]
+1 from me.

**Comment by jhpalmieri** [comment]
+1. Looks fine to me.

**Comment by mkoeppe** [comment]
Thanks @roed314 @jhpalmieri.

This brings the tally to 4-2, setting to "positive review".

**Pull request merged**